clash-lib/src/Clash/Core/Evaluator/Models.hs

```diff
@@ -1,202 +1,274 @@
-{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
 {-# LANGUAGE LambdaCase #-}
 {-# LANGUAGE OverloadedStrings #-}
 
 module Clash.Core.Evaluator.Models where
 
 import Control.Concurrent.Supply (Supply)
-import Control.Monad ((>=>))
-import Control.Monad.State.Strict (State)
-import Control.DeepSeq (NFData(..))
+import Control.Monad.RWS.Strict (MonadReader, MonadState, RWS)
+import qualified Control.Monad.RWS.Strict as RWS
```

Comment:
I'm not sure about this, the CPS variant is only to make the Writer part constant space (which we don't use) and would incur extra boilerplate (`mtl` doesn't provide instances for the CPS variant yet).

Would you be happy with a comment saying that if the Writer part is used in the future to change to the CPS module instead?

Reply:
But couldn't you use the `transformers` functions instead of the `mtl` one?:
* http://hackage.haskell.org/package/transformers-0.5.6.2/docs/Control-Monad-Trans-RWS-CPS.html#g:3
* http://hackage.haskell.org/package/transformers-0.5.6.2/docs/Control-Monad-Trans-RWS-CPS.html#g:5

Unless we're using some code that has a `MonadState` and `MonadReader` constraint of course and we have to use `mtl`.

Reply:
Currently the code uses the `mtl` instances derived for `Eval`, changing to the `transformers` ones directly would only work if `Eval` was changed from `newtype` to `type` or the `MonadState` and `MonadReader` classes were implemented for `Eval` using the functions in the `CPS` module.

Given we don't use the `Writer`, I would rather keep it as-is, and change the `mtl` import when the next version of `mtl` is released (which includes [a new CPS module](https://github.com/haskell/mtl/blob/master/Control/Monad/RWS/CPS.hs))

Reply:
Alright, sounds good.